src/common/base/abstractQuick.ts

```diff
@@ -0,0 +1,171 @@
+import { async } from '../flags/promote/promoteFlags';
+import { concise, requiredDoceOrgFlag, verbose, wait } from '../flags/flags';
+import { OutputServiceFactory } from '../outputService';
+import { AsyncOperationResultJson } from '../../common';
+import {
+  CheckDeploymentResultWithChangeBundleInstalls,
+  selectOneDeploymentResultWithChangeBundleInstallsByAsyncJobId,
+} from '../selectors/deploymentResultsSelector';
+import { selectOnePipelineStageByEnvironmentId } from '../selectors/pipelineStageSelector';
+import { AsyncCommand } from './abstractAsyncOperation';
+
+export type Flags<T extends typeof SfCommand> = Interfaces.InferredFlags<
+  (typeof QuickPromotionCommand)['baseFlags'] & T['flags']
+>;
+
+Messages.importMessagesDirectory(__dirname);
+const messages = Messages.loadMessages('@salesforce/plugin-devops-center', 'project.deploy.pipeline.quick');
+
+export default abstract class QuickPromotionCommand<T extends typeof SfCommand> extends AsyncCommand {
+  public static baseFlags = {
+    async,
+    concise,
+    verbose,
+    wait,
+    'devops-center-username': requiredDoceOrgFlag(),
+    'job-id': Flags.salesforceId({
+      char: 'i',
+      description: messages.getMessage('flags.job-id.description'),
+      summary: messages.getMessage('flags.job-id.summary'),
+      exactlyOne: ['use-most-recent', 'job-id'],
+    }),
+    'use-most-recent': Flags.boolean({
+      char: 'r',
+      description: messages.getMessage('flags.use-most-recent.description'),
+      summary: messages.getMessage('flags.use-most-recent.summary'),
+      exactlyOne: ['use-most-recent', 'job-id'],
+    }),
+  };
+
+  protected flags!: Flags<T>;
+  private targetStage: PipelineStage;
+  private deployOptions: Partial<PromoteOptions>;
+
+  public async init(): Promise<void> {
+    await super.init();
+    const { flags } = await this.parse({
+      flags: this.ctor.flags,
+      baseFlags: (super.ctor as typeof QuickPromotionCommand).baseFlags,
+    });
+    this.flags = flags as Flags<T>;
+    this.targetOrg = this.flags['devops-center-username'] as Org;
+  }
+
+  /**
+   *
+   * Knows how to perform a quick promotion and watch its progress.
+   */
+  public async executeQuickPromotion(): Promise<AsyncOperationResultJson> {
+    const deploymentResult = await this.getDeploymentIdForQuickDeploy();
+    const envId = deploymentResult.sf_devops__Change_Bundle_Installs__r.records[0].sf_devops__Environment__c;
+    this.targetStage = (await selectOnePipelineStageByEnvironmentId(
+      this.targetOrg.getConnection(),
+      envId
+    )) as PipelineStage;
+
+    this.setOutputService(
+      new OutputServiceFactory().forQuickDeployment(
+        this.flags,
+        (this.flags['devops-center-username'] as Org).getConnection(),
+        this.targetStage.sf_devops__Branch__r.sf_devops__Name__c
+      )
+    );
+
+    this.setAsyncOperationId(
+      await this.requestPromotion(
+        this.targetStage?.sf_devops__Pipeline__r.sf_devops__Project__c,
```

Comment:
This can be null/undefined ???

Reply:
Nop.

Reply:
Why we have the **?** ?

Reply:
It's not needed, I will remove it. Thanks.